flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.parquet;
+
+import org.apache.flink.api.common.io.CheckpointableInputFormat;
+import org.apache.flink.api.common.io.FileInputFormat;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.core.fs.FileInputSplit;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.formats.parquet.utils.ParquetRecordReader;
+import org.apache.flink.formats.parquet.utils.ParquetSchemaConverter;
+import org.apache.flink.formats.parquet.utils.RowReadSupport;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.parquet.ParquetReadOptions;
+import org.apache.parquet.filter2.compat.FilterCompat;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.util.HadoopInputFile;
+import org.apache.parquet.io.InputFile;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The base InputFormat class to read from Parquet files.
+ * For specific return types the {@link #convert(Row)} method need to be implemented.
+ *
+ * <P>Using {@link ParquetRecordReader} to Read files instead of {@link org.apache.flink.core.fs.FSDataInputStream},
+ * we override {@link #open(FileInputSplit)} and {@link #close()} to change the behaviors.
+ *
+ * @param <E> The type of record to read.
+ */
+public abstract class ParquetInputFormat<E> extends FileInputFormat<E> implements
+	CheckpointableInputFormat<FileInputSplit, Tuple2<Long, Long>> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(ParquetInputFormat.class);
+
+	private transient Counter recordConsumed;
+
+	protected RowTypeInfo readType;
+
+	protected boolean isStandard;
+
+	protected final TypeInformation[] fieldTypes;
+
+	protected final String[] fieldNames;
+
+	protected transient ParquetRecordReader<Row> parquetRecordReader;
+
+	protected transient long recordsReadSinceLastSync;
+
+	protected long lastSyncedBlock = -1L;
+
+	protected ParquetInputFormat(Path path, TypeInformation[] fieldTypes, String[] fieldNames, boolean isStandard) {
+		super(path);
+		this.readType = new RowTypeInfo(fieldTypes, fieldNames);
+		this.fieldTypes = readType.getFieldTypes();
+		this.fieldNames = readType.getFieldNames();
+		this.unsplittable = true;
```

Comment:
As a Parquet file is composed of row groups with a separate metadata in footer, it is definitely splittable. It acquires to read footer of each files in advance and process each file slit with the offset info of original file. It could be separate improvement that can be done in a separate PR. How do you think?

Reply:
Hey @HuangZhenQiu, did you end up filing a JIRA for adding rowgroup-level splitting? Couldn't find it linked from the PR ticket.

Reply:
@a10y 
This PR is closed and merged into master. I am working on the ParquetTableSource with predicate pushdown. After that, Another PR will be added for rowgroup level splitting.